File: python/dcc_mcp_core/script_materialization.py
"""Per-instance script materialization store (#1220).

The store writes ad-hoc script content to a host-local file before execution.
It keeps the path contract tied to the DCC type, live instance, and MCP
session so downstream execution, audit, sandbox allowlists, and cleanup can
reason about a concrete file instead of an opaque inline ``code`` string.
"""

from __future__ import annotations

import contextlib
from dataclasses import dataclass
from datetime import datetime
from datetime import timedelta
from datetime import timezone
import hashlib
import json
import os
from pathlib import Path
import re
import tempfile
from typing import Any
from typing import Mapping
import uuid

SCRIPT_MATERIALIZATION_ROOT_ENV = "DCC_MCP_SCRIPT_MATERIALIZATION_ROOT"
DEFAULT_SCRIPT_MATERIALIZATION_ROOT = Path.home() / ".dcc-mcp"

_SEGMENT_RE = re.compile(r"[^A-Za-z0-9_.-]+")
_SAFE_SUFFIX_RE = re.compile(r"[^A-Za-z0-9_.-]+")


@dataclass(frozen=True)
class MaterializedScript:
    """Structured descriptor returned by :func:`materialize_script`."""

    file_ref: dict[str, Any]
    file_path: str
    path: str
    language: str
    suffix: str
    sha256: str
    bytes: int
    created_at: str
    expires_at: str | None
    ttl_secs: int | None
    dcc_type: str
    instance_id: str
    session_id: str
    script_id: str
    tool_call_id: str | None = None
    correlation_id: str | None = None
    reused: bool = False

    @property
    def file_ref_uri(self) -> str:
        """Return the FileRef-compatible URI for this script."""
        return str(self.file_ref["uri"])

    def to_dict(self) -> dict[str, Any]:
        """Return a JSON-serializable descriptor."""
        data = {
            "file_ref": self.file_ref,
            "file_ref_uri": self.file_ref_uri,
            "file_path": self.file_path,
            "path": self.path,
            "language": self.language,
            "suffix": self.suffix,
            "sha256": self.sha256,
            "bytes": self.bytes,
            "created_at": self.created_at,
            "expires_at": self.expires_at,
            "ttl_secs": self.ttl_secs,
            "dcc_type": self.dcc_type,
            "instance_id": self.instance_id,
            "session_id": self.session_id,
            "script_id": self.script_id,
            "tool_call_id": self.tool_call_id,
            "correlation_id": self.correlation_id,
            "reused": self.reused,
        }
        return {key: value for key, value in data.items() if value is not None}


def default_script_materialization_root(root: str | os.PathLike[str] | None = None) -> Path:
    """Return the configured script materialization root.

    ``root`` wins first, then ``DCC_MCP_SCRIPT_MATERIALIZATION_ROOT``, then the
    documented default ``~/.dcc-mcp``.
    """
    if root is not None:
        return Path(root).expanduser()
    configured = os.environ.get(SCRIPT_MATERIALIZATION_ROOT_ENV)
    if configured:
        return Path(configured).expanduser()
    return DEFAULT_SCRIPT_MATERIALIZATION_ROOT


def sanitize_materialization_segment(value: str | None, *, default: str = "unknown") -> str:
    """Return a single filesystem-safe path segment.

    Separators, drive delimiters, traversal tokens, whitespace, and other
    special characters collapse to ``_``. Custom DCC names remain readable
    when they already use safe ASCII characters.
    """
    raw = "" if value is None else str(value).strip()
    raw = raw.replace("\\", "_").replace("/", "_").replace(":", "_")
    safe = _SEGMENT_RE.sub("_", raw).strip("._-")
    if not safe or safe in {".", ".."}:
        return default
    return safe[:96]


def _normalize_suffix(suffix: str | None) -> str:
    raw = ".py" if not suffix else str(suffix).strip()
    raw = raw.replace("\\", "_").replace("/", "_").replace(":", "_")
    raw = raw.lstrip(".")
    safe = "." + _SAFE_SUFFIX_RE.sub("_", raw)
    if safe in {".", ".."}:
        return ".txt"
    return safe[:32]


def _utc_now() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def _format_utc(value: datetime) -> str:
    return value.replace(microsecond=0).isoformat() + "Z"


def _parse_utc(value: str | None) -> datetime | None:
    if not value:
        return None
    text = value[:-1] if value.endswith("Z") else value
    try:
        return datetime.strptime(text, "%Y-%m-%dT%H:%M:%S")
    except ValueError:
        return None


def _resolve_path(path: Path) -> Path:
    return path.resolve()


def _ensure_within_root(path: Path, root: Path) -> Path:
    resolved_root = _resolve_path(root)
    resolved_path = _resolve_path(path)
    try:
        resolved_path.relative_to(resolved_root)
    except ValueError as exc:
        raise ValueError(f"materialized script path escapes root: {path}") from exc
    return resolved_path


def _lexical_path_under_root(path: Path, root: Path) -> bool:
    parent = path.parent.resolve() if path.parent.exists() else path.parent
    candidate = parent / path.name
    try:
        candidate.relative_to(root)
    except ValueError:
        return False
    return True


def _atomic_write(path: Path, data: bytes) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp_name = tempfile.mkstemp(
        prefix=".tmp-",
        suffix=".part",
        dir=str(path.parent),
    )
    tmp = Path(tmp_name)
    try:
        with os.fdopen(fd, "wb") as fp:
            fp.write(data)
            fp.flush()
            os.fsync(fp.fileno())
        tmp.replace(path)
    except Exception:
        with contextlib.suppress(Exception):
            tmp.unlink()
        raise


def _metadata_path(script_path: Path) -> Path:
    return script_path.with_name(script_path.name + ".meta.json")


def _mime_for_language(language: str) -> str:
    normalized = language.lower()
    if normalized in {"python", "py"}:
        return "text/x-python"
    if normalized in {"mel"}:
        return "text/x-mel"
    if normalized in {"javascript", "js"}:
        return "text/javascript"
    if normalized in {"powershell", "ps1"}:
        return "text/x-powershell"
    return "text/plain"


def _file_ref_from_descriptor(descriptor: Mapping[str, Any]) -> dict[str, Any]:
    metadata = {
        "dcc_type": descriptor["dcc_type"],
        "instance_id": descriptor["instance_id"],
        "session_id": descriptor["session_id"],
        "script_id": descriptor["script_id"],
        "language": descriptor["language"],
        "suffix": descriptor["suffix"],
        "materialization_kind": "script",
    }
    return {
        "uri": Path(str(descriptor["file_path"])).as_uri(),
        "mime": _mime_for_language(str(descriptor["language"])),
        "size_bytes": descriptor["bytes"],
        "display_name": Path(str(descriptor["file_path"])).name,
        "digest": f"sha256:{descriptor['sha256']}",
        "tool_call_id": descriptor.get("tool_call_id"),
        "session_id": descriptor["session_id"],
        "correlation_id": descriptor.get("correlation_id"),
        "created_at": descriptor["created_at"],
        "expires_at": descriptor.get("expires_at"),
        "metadata": metadata,
    }


def _descriptor_from_dict(data: Mapping[str, Any], *, reused: bool) -> MaterializedScript:
    file_ref = data.get("file_ref")
    if not isinstance(file_ref, dict):
        file_ref = _file_ref_from_descriptor(data)
    return MaterializedScript(
        file_ref=dict(file_ref),
        file_path=str(data["file_path"]),
        path=str(data.get("path", data["file_path"])),
        language=str(data["language"]),
        suffix=str(data["suffix"]),
        sha256=str(data["sha256"]),
        bytes=int(data["bytes"]),
        created_at=str(data["created_at"]),
        expires_at=data.get("expires_at"),
        ttl_secs=data.get("ttl_secs"),
        dcc_type=str(data["dcc_type"]),
        instance_id=str(data["instance_id"]),
        session_id=str(data["session_id"]),
        script_id=str(data["script_id"]),
        tool_call_id=data.get("tool_call_id"),
        correlation_id=data.get("correlation_id"),
        reused=reused,
    )


def _script_id_for(
    *,
    sha256: str,
    reuse: bool,
    reuse_key: str | None,
    prefix: str,
) -> str:
    if reuse:
        if reuse_key:
            key = sanitize_materialization_segment(reuse_key, default="reuse")
            return f"{key}_{sha256[:12]}"
        return sha256
    safe_prefix = sanitize_materialization_segment(prefix, default="script")
    return f"{safe_prefix}_{uuid.uuid4().hex}_{sha256[:12]}"


def materialize_script(
    content: str,
    *,
    dcc_type: str,
    instance_id: str,
    session_id: str,
    language: str = "python",
    suffix: str = ".py",
    display_name: str | None = None,
    reuse: bool = False,
    reuse_key: str | None = None,
    ttl_secs: int | None = None,
    root: str | os.PathLike[str] | None = None,
    tool_call_id: str | None = None,
    correlation_id: str | None = None,
    prefix: str = "script",
) -> MaterializedScript:
    """Materialize script content under a DCC/session-scoped host path."""
    if not isinstance(content, str):
        raise TypeError("content must be a string")
    if ttl_secs is not None and (isinstance(ttl_secs, bool) or ttl_secs <= 0):
        raise ValueError("ttl_secs must be a positive integer")

    body = content.encode("utf-8")
    sha256 = hashlib.sha256(body).hexdigest()
    suffix = _normalize_suffix(suffix)
    language = sanitize_materialization_segment(language, default="text").lower()
    safe_dcc = sanitize_materialization_segment(dcc_type)
    safe_instance = sanitize_materialization_segment(instance_id)
    safe_session = sanitize_materialization_segment(session_id)
    script_id = _script_id_for(
        sha256=sha256,
        reuse=reuse,
        reuse_key=reuse_key,
        prefix=display_name or prefix,
    )

    store_root = default_script_materialization_root(root)
    target_dir = store_root / safe_dcc / "temp" / safe_instance / safe_session
    target_path = target_dir / f"{script_id}{suffix}"
    target_dir.mkdir(parents=True, exist_ok=True)
    resolved_target = _ensure_within_root(target_path, store_root)
    metadata_path = _metadata_path(resolved_target)

    if reuse and resolved_target.is_file() and metadata_path.is_file():
        try:
            existing = json.loads(metadata_path.read_text(encoding="utf-8"))
        except (OSError, ValueError, TypeError):
            existing = None
        if isinstance(existing, dict) and existing.get("sha256") == sha256:
            expires_at = _parse_utc(existing.get("expires_at"))
            if expires_at is None or expires_at > _utc_now():
                return _descriptor_from_dict(existing, reused=True)

    now = _utc_now()
    expires_at = _format_utc(now + timedelta(seconds=ttl_secs)) if ttl_secs else None
    descriptor = {
        "file_path": str(resolved_target),
        "path": str(resolved_target),
        "language": language,
        "suffix": suffix,
        "sha256": sha256,
        "bytes": len(body),
        "created_at": _format_utc(now),
        "expires_at": expires_at,
        "ttl_secs": ttl_secs,
        "dcc_type": safe_dcc,
        "instance_id": safe_instance,
        "session_id": safe_session,
        "script_id": script_id,
        "tool_call_id": tool_call_id,
        "correlation_id": correlation_id,
        "reused": False,
    }
    descriptor["file_ref"] = _file_ref_from_descriptor(descriptor)

    _atomic_write(resolved_target, body)
    _atomic_write(
        metadata_path,
        json.dumps(descriptor, sort_keys=True, separators=(",", ":")).encode("utf-8"),
    )
    return _descriptor_from_dict(descriptor, reused=False)


def cleanup_materialized_scripts(
    *,
    root: str | os.PathLike[str] | None = None,
    now: datetime | None = None,
    include_unexpired: bool = False,
) -> int:
    """Remove expired script files and sidecars below ``root``.

    The cleanup pass only unlinks files whose metadata sidecar lives below the
    configured materialization root. It does not recursively remove arbitrary
    directories and it does not follow symlink targets.
    """
    store_root = default_script_materialization_root(root)
    if not store_root.exists():
        return 0

    resolved_root = _resolve_path(store_root)
    cutoff = now or _utc_now()
    removed = 0
    for sidecar in resolved_root.rglob("*.meta.json"):
        with contextlib.suppress(ValueError):
            sidecar.relative_to(resolved_root)
        try:
            data = json.loads(sidecar.read_text(encoding="utf-8"))
        except (OSError, ValueError, TypeError):
            continue
        expires_at = _parse_utc(data.get("expires_at") if isinstance(data, dict) else None)
        if not include_unexpired and (expires_at is None or expires_at > cutoff):
            continue
        script_path = Path(str(data.get("file_path", ""))) if isinstance(data, dict) else Path()
        targets = [script_path, sidecar]
        for target in targets:
            with contextlib.suppress(Exception):
                if _lexical_path_under_root(target, resolved_root) and (target.is_file() or target.is_symlink()):
                    target.unlink()
                    removed += 1
    return removed


__all__ = [
    "DEFAULT_SCRIPT_MATERIALIZATION_ROOT",
    "SCRIPT_MATERIALIZATION_ROOT_ENV",
    "MaterializedScript",
    "cleanup_materialized_scripts",
    "default_script_materialization_root",
    "materialize_script",
    "sanitize_materialization_segment",
]

File: crates/dcc-mcp-artefact/src/script_materialization.rs
//! Per-instance script materialization store (#1220).
//!
//! The store writes ad-hoc script content to a host-local path before a DCC
//! interpreter runs it. The descriptor preserves the DCC type, live instance,
//! MCP session, request metadata, and a [`FileRef`] compatible URI so audit,
//! replay, and sandbox allowlists can reason about a concrete file.

use std::fs;
use std::io;
use std::path::{Path, PathBuf};

use chrono::{DateTime, Duration, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{FileRef, atomic_write_bytes, hash_bytes_sha256};

/// Environment override for the default materialization root.
pub const SCRIPT_MATERIALIZATION_ROOT_ENV: &str = "DCC_MCP_SCRIPT_MATERIALIZATION_ROOT";

/// Request for materializing one script payload.
#[derive(Debug, Clone)]
pub struct ScriptMaterializationRequest {
    pub content: String,
    pub dcc_type: String,
    pub instance_id: String,
    pub session_id: String,
    pub language: String,
    pub suffix: String,
    pub display_name: Option<String>,
    pub reuse: bool,
    pub reuse_key: Option<String>,
    pub ttl_secs: Option<u64>,
    pub tool_call_id: Option<String>,
    pub correlation_id: Option<String>,
}

impl ScriptMaterializationRequest {
    /// Build a request with Python-friendly defaults.
    pub fn new(
        content: impl Into<String>,
        dcc_type: impl Into<String>,
        instance_id: impl Into<String>,
        session_id: impl Into<String>,
    ) -> Self {
        Self {
            content: content.into(),
            dcc_type: dcc_type.into(),
            instance_id: instance_id.into(),
            session_id: session_id.into(),
            language: "python".to_string(),
            suffix: ".py".to_string(),
            display_name: None,
            reuse: false,
            reuse_key: None,
            ttl_secs: None,
            tool_call_id: None,
            correlation_id: None,
        }
    }
}

/// Structured descriptor returned by [`ScriptMaterializationStore`].
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct MaterializedScript {
    pub file_ref: FileRef,
    pub file_path: PathBuf,
    pub sha256: String,
    pub bytes: u64,
    pub language: String,
    pub suffix: String,
    pub dcc_type: String,
    pub instance_id: String,
    pub session_id: String,
    pub script_id: String,
    pub ttl_secs: Option<u64>,
    pub created_at: DateTime<Utc>,
    pub expires_at: Option<DateTime<Utc>>,
    pub tool_call_id: Option<String>,
    pub correlation_id: Option<String>,
    pub reused: bool,
}

impl MaterializedScript {
    /// Return true when this descriptor is expired at ``now``.
    pub fn is_expired_at(&self, now: DateTime<Utc>) -> bool {
        self.expires_at.is_some_and(|expires_at| expires_at <= now)
    }
}

/// Errors returned by script materialization.
#[must_use]
#[derive(Debug, thiserror::Error)]
pub enum ScriptMaterializationError {
    #[error("invalid materialization request: {0}")]
    InvalidRequest(String),
    #[error("materialized script path escapes root: {0}")]
    PathEscape(String),
    #[error("I/O error: {0}")]
    Io(#[from] io::Error),
    #[error("serialization error: {0}")]
    Serde(#[from] serde_json::Error),
}

pub type ScriptMaterializationResult<T> = Result<T, ScriptMaterializationError>;

/// Store rooted at ``~/.dcc-mcp`` or ``DCC_MCP_SCRIPT_MATERIALIZATION_ROOT``.
#[derive(Debug, Clone)]
pub struct ScriptMaterializationStore {
    root: PathBuf,
}

impl ScriptMaterializationStore {
    /// Create a store under an explicit root.
    pub fn new_in(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }

    /// Create a store under the configured default root.
    pub fn new_default() -> Self {
        Self::new_in(default_script_materialization_root())
    }

    /// Return the configured root.
    pub fn root(&self) -> &Path {
        &self.root
    }

    /// Materialize one script and return its descriptor.
    pub fn materialize(
        &self,
        request: ScriptMaterializationRequest,
    ) -> ScriptMaterializationResult<MaterializedScript> {
        validate_ttl(request.ttl_secs)?;
        let bytes = request.content.as_bytes();
        let sha256 = hash_bytes_sha256(bytes);
        let suffix = normalize_suffix(&request.suffix);
        let language = sanitize_segment(&request.language, "text").to_ascii_lowercase();
        let dcc_type = sanitize_segment(&request.dcc_type, "unknown");
        let instance_id = sanitize_segment(&request.instance_id, "unknown");
        let session_id = sanitize_segment(&request.session_id, "unknown");
        let script_id = script_id_for(
            &sha256,
            request.reuse,
            request.reuse_key.as_deref(),
            request.display_name.as_deref().unwrap_or("script"),
        );

        let target_dir = self
            .root
            .join(&dcc_type)
            .join("temp")
            .join(&instance_id)
            .join(&session_id);
        fs::create_dir_all(&target_dir)?;
        ensure_within_root(&target_dir, &self.root)?;
        let file_path = target_dir.join(format!("{script_id}{suffix}"));
        ensure_within_root(&file_path, &self.root)?;
        let metadata_path = metadata_path_for(&file_path);

        if request.reuse
            && file_path.is_file()
            && metadata_path.is_file()
            && let Some(existing) = read_descriptor_lenient(&metadata_path)?
            && existing.sha256 == sha256
            && !existing.is_expired_at(Utc::now())
        {
            return Ok(MaterializedScript {
                reused: true,
                ..existing
            });
        }

        let now = Utc::now();
        let expires_at = expires_at_from_ttl(request.ttl_secs);
        let file_ref = FileRef {
            uri: file_uri_from_path(&file_path),
            mime: Some(mime_for_language(&language).to_string()),
            size_bytes: Some(bytes.len() as u64),
            display_name: request.display_name.clone().or_else(|| {
                file_path
                    .file_name()
                    .map(|name| name.to_string_lossy().into_owned())
            }),
            digest: Some(format!("sha256:{sha256}")),
            producer_job_id: None,
            tool_call_id: request.tool_call_id.clone(),
            session_id: Some(session_id.clone()),
            correlation_id: request.correlation_id.clone(),
            created_at: now,
            expires_at,
            metadata: serde_json::json!({
                "materialization_kind": "script",
                "dcc_type": dcc_type,
                "instance_id": instance_id,
                "session_id": session_id,
                "script_id": script_id,
                "language": language,
                "suffix": suffix,
            }),
        };
        let descriptor = MaterializedScript {
            file_ref,
            file_path,
            sha256,
            bytes: bytes.len() as u64,
            language,
            suffix,
            dcc_type,
            instance_id,
            session_id,
            script_id,
            ttl_secs: request.ttl_secs,
            created_at: now,
            expires_at,
            tool_call_id: request.tool_call_id,
            correlation_id: request.correlation_id,
            reused: false,
        };

        atomic_write_bytes(&descriptor.file_path, bytes, true)?;
        atomic_write_bytes(
            &metadata_path,
            &serde_json::to_vec(&descriptor).map_err(ScriptMaterializationError::Serde)?,
            true,
        )?;
        Ok(descriptor)
    }

    /// Remove expired scripts and metadata sidecars. Returns removed file count.
    pub fn cleanup_expired(&self, now: DateTime<Utc>) -> ScriptMaterializationResult<usize> {
        cleanup_materialized_scripts(&self.root, now, false)
    }

    /// Remove all scripts and metadata sidecars below the store root.
    pub fn cleanup_all(&self) -> ScriptMaterializationResult<usize> {
        cleanup_materialized_scripts(&self.root, Utc::now(), true)
    }
}

/// Return the default materialization root.
pub fn default_script_materialization_root() -> PathBuf {
    if let Some(root) = std::env::var_os(SCRIPT_MATERIALIZATION_ROOT_ENV) {
        return PathBuf::from(root);
    }
    home_dir()
        .unwrap_or_else(std::env::temp_dir)
        .join(".dcc-mcp")
}

/// Sanitize one path segment for DCC, instance, session, and script ids.
pub fn sanitize_segment(value: &str, default: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for ch in value.trim().chars() {
        if ch.is_ascii_alphanumeric() || matches!(ch, '_' | '-' | '.') {
            out.push(ch);
        } else {
            out.push('_');
        }
    }
    let trimmed = out.trim_matches(['.', '_', '-']).to_string();
    if trimmed.is_empty() || trimmed == "." || trimmed == ".." {
        return default.to_string();
    }
    trimmed.chars().take(96).collect()
}

fn home_dir() -> Option<PathBuf> {
    std::env::var_os("HOME")
        .or_else(|| std::env::var_os("USERPROFILE"))
        .map(PathBuf::from)
}

fn validate_ttl(ttl_secs: Option<u64>) -> ScriptMaterializationResult<()> {
    if ttl_secs == Some(0) {
        return Err(ScriptMaterializationError::InvalidRequest(
            "ttl_secs must be greater than zero".to_string(),
        ));
    }
    Ok(())
}

fn normalize_suffix(suffix: &str) -> String {
    let raw = if suffix.trim().is_empty() {
        ".py"
    } else {
        suffix.trim()
    };
    let mut out = String::new();
    out.push('.');
    let trimmed = raw.trim_start_matches('.');
    for ch in trimmed.chars() {
        if ch.is_ascii_alphanumeric() || matches!(ch, '_' | '-' | '.') {
            out.push(ch);
        } else {
            out.push('_');
        }
    }
    if out == "." || out == ".." {
        return ".txt".to_string();
    }
    out.chars().take(32).collect()
}

fn script_id_for(sha256: &str, reuse: bool, reuse_key: Option<&str>, prefix: &str) -> String {
    if reuse {
        if let Some(reuse_key) = reuse_key {
            return format!("{}_{}", sanitize_segment(reuse_key, "reuse"), &sha256[..12]);
        }
        return sha256.to_string();
    }
    format!(
        "{}_{}_{}",
        sanitize_segment(prefix, "script"),
        Uuid::new_v4().simple(),
        &sha256[..12]
    )
}

fn ensure_within_root(path: &Path, root: &Path) -> ScriptMaterializationResult<()> {
    let root = canonicalize_existing_or_parent(root)?;
    let path = canonicalize_existing_or_parent(path)?;
    if !path.starts_with(&root) {
        return Err(ScriptMaterializationError::PathEscape(
            path.to_string_lossy().into_owned(),
        ));
    }
    Ok(())
}

fn canonicalize_existing_or_parent(path: &Path) -> io::Result<PathBuf> {
    if path.exists() {
        return path.canonicalize();
    }
    if let Some(parent) = path.parent()
        && parent.exists()
    {
        return Ok(parent
            .canonicalize()?
            .join(path.file_name().unwrap_or_default()));
    }
    Ok(path.to_path_buf())
}

fn metadata_path_for(path: &Path) -> PathBuf {
    let name = path
        .file_name()
        .map(|name| name.to_string_lossy().into_owned())
        .unwrap_or_else(|| "script".to_string());
    path.with_file_name(format!("{name}.meta.json"))
}

fn read_descriptor(path: &Path) -> ScriptMaterializationResult<Option<MaterializedScript>> {
    match fs::read(path) {
        Ok(bytes) => Ok(Some(serde_json::from_slice(&bytes)?)),
        Err(err) if err.kind() == io::ErrorKind::NotFound => Ok(None),
        Err(err) => Err(err.into()),
    }
}

fn read_descriptor_lenient(path: &Path) -> ScriptMaterializationResult<Option<MaterializedScript>> {
    match read_descriptor(path) {
        Ok(descriptor) => Ok(descriptor),
        Err(ScriptMaterializationError::Serde(_)) => Ok(None),
        Err(err) => Err(err),
    }
}

fn expires_at_from_ttl(ttl_secs: Option<u64>) -> Option<DateTime<Utc>> {
    let ttl = i64::try_from(ttl_secs?).ok()?;
    Utc::now().checked_add_signed(Duration::seconds(ttl))
}

fn file_uri_from_path(path: &Path) -> String {
    let text = path.to_string_lossy().replace('\\', "/");
    if text.starts_with('/') {
        format!("file://{text}")
    } else {
        format!("file:///{text}")
    }
}

fn mime_for_language(language: &str) -> &'static str {
    match language {
        "python" | "py" => "text/x-python",
        "mel" => "text/x-mel",
        "javascript" | "js" => "text/javascript",
        "powershell" | "ps1" => "text/x-powershell",
        _ => "text/plain",
    }
}

fn cleanup_materialized_scripts(
    root: &Path,
    now: DateTime<Utc>,
    include_unexpired: bool,
) -> ScriptMaterializationResult<usize> {
    if !root.exists() {
        return Ok(0);
    }
    let root = root.canonicalize()?;
    let mut removed = 0;
    let mut stack = vec![root.clone()];
    while let Some(dir) = stack.pop() {
        for entry in fs::read_dir(&dir)? {
            let entry = entry?;
            let file_type = entry.file_type()?;
            let path = entry.path();
            if file_type.is_dir() {
                stack.push(path);
                continue;
            }
            if path.extension().and_then(|ext| ext.to_str()) != Some("json")
                || !path
                    .file_name()
                    .and_then(|name| name.to_str())
                    .is_some_and(|name| name.ends_with(".meta.json"))
            {
                continue;
            }
            let Some(descriptor) = read_descriptor_lenient(&path)? else {
                continue;
            };
            if !include_unexpired && !descriptor.is_expired_at(now) {
                continue;
            }
            if remove_file_or_symlink_under_root(&descriptor.file_path, &root)? {
                removed += 1;
            }
            if remove_file_or_symlink_under_root(&path, &root)? {
                removed += 1;
            }
        }
    }
    Ok(removed)
}

fn remove_file_or_symlink_under_root(path: &Path, root: &Path) -> io::Result<bool> {
    let candidate = canonicalize_parent_join(path)?;
    if !candidate.starts_with(root) {
        return Ok(false);
    }
    let metadata = match fs::symlink_metadata(path) {
        Ok(metadata) => metadata,
        Err(err) if err.kind() == io::ErrorKind::NotFound => return Ok(false),
        Err(err) => return Err(err),
    };
    let file_type = metadata.file_type();
    if file_type.is_file() || file_type.is_symlink() {
        fs::remove_file(path)?;
        return Ok(true);
    }
    Ok(false)
}

fn canonicalize_parent_join(path: &Path) -> io::Result<PathBuf> {
    if let Some(parent) = path.parent()
        && parent.exists()
    {
        return Ok(parent
            .canonicalize()?
            .join(path.file_name().unwrap_or_default()));
    }
    Ok(path.to_path_buf())
}

#[cfg(test)]
mod tests {
    use std::thread;
    use std::time::Duration as StdDuration;

    use tempfile::tempdir;

    use super::*;

    #[test]
    fn sanitizes_path_segments_and_windows_style_values() {
        assert_eq!(sanitize_segment("maya/../../prod", "x"), "maya_.._.._prod");
        assert_eq!(
            sanitize_segment("C:\\show\\maya:2026", "x"),
            "C__show_maya_2026"
        );
        assert_eq!(sanitize_segment("...", "fallback"), "fallback");
    }

    #[test]
    fn materializes_descriptor_with_file_ref_metadata() {
        let tmp = tempdir().unwrap();
        let store = ScriptMaterializationStore::new_in(tmp.path());
        let mut request =
            ScriptMaterializationRequest::new("print('hello')", "maya", "inst-1", "sess-1");
        request.tool_call_id = Some("tool-1".to_string());
        request.correlation_id = Some("corr-1".to_string());
        request.ttl_secs = Some(60);

        let descriptor = store.materialize(request).unwrap();

        assert!(descriptor.file_path.is_file());
        assert_eq!(descriptor.bytes, 14);
        assert_eq!(descriptor.file_ref.session_id.as_deref(), Some("sess-1"));
        assert_eq!(descriptor.file_ref.tool_call_id.as_deref(), Some("tool-1"));
        assert_eq!(
            descriptor.file_ref.digest,
            Some(format!("sha256:{}", descriptor.sha256))
        );
        assert_eq!(descriptor.file_ref.metadata["dcc_type"], "maya");
        assert!(descriptor.file_ref.uri.starts_with("file://"));
    }

    #[test]
    fn reuses_identical_content_when_requested() {
        let tmp = tempdir().unwrap();
        let store = ScriptMaterializationStore::new_in(tmp.path());
        let mut request = ScriptMaterializationRequest::new("x = 1", "custom", "i", "s");
        request.reuse = true;
        request.reuse_key = Some("bootstrap".to_string());

        let first = store.materialize(request.clone()).unwrap();
        let second = store.materialize(request).unwrap();

        assert_eq!(first.file_path, second.file_path);
        assert!(!first.reused);
        assert!(second.reused);
    }

    #[test]
    fn cleanup_removes_expired_scripts_only() {
        let tmp = tempdir().unwrap();
        let store = ScriptMaterializationStore::new_in(tmp.path());
        let mut expiring = ScriptMaterializationRequest::new("x = 1", "maya", "i", "s");
        expiring.ttl_secs = Some(1);
        let expired = store.materialize(expiring).unwrap();
        let live = store
            .materialize(ScriptMaterializationRequest::new("x = 2", "maya", "i", "s"))
            .unwrap();

        thread::sleep(StdDuration::from_secs(2));
        let removed = store.cleanup_expired(Utc::now()).unwrap();

        assert!(removed >= 2);
        assert!(!expired.file_path.exists());
        assert!(live.file_path.exists());
    }
}
